src/pages/api/webhooks/image-resource-training.ts

```diff
@@ -0,0 +1,133 @@
+import { Prisma, TrainingStatus } from '@prisma/client';
+import * as z from 'zod';
+import { dbWrite } from '~/server/db/client';
+import { modelFileMetadataSchema } from '~/server/schema/model-file.schema';
+import { WebhookEndpoint } from '~/server/utils/endpoint-helpers';
+
+export type EpochSchema = z.infer<typeof epoch_schema>;
+const epoch_schema = z.object({
+  epoch_number: z.number(),
+  model_url: z.string(),
+  sample_images: z
+    .array(
+      z.object({
+        image_url: z.string(),
+        prompt: z.string(),
+      })
+    )
+    .optional(),
+});
+
+type ContextProps = z.infer<typeof context>;
+const context = z.object({
+  modelFileId: z.number(),
+  status: z.string(),
+  message: z.string().optional(),
+  model: z.string(),
+  start_time: z.number(),
+  end_time: z.number().optional(),
+  duration: z.number().optional(),
+  epochs: z.array(epoch_schema).optional(),
+  upload_duration: z.number().optional(),
+  sample_prompts: z.array(z.string()).optional(),
+  logs: z
+    .object({
+      stdout: z.string().optional(),
+      stderr: z.string().optional(),
+    })
+    .optional(),
+});
+
+const schema = z.object({
+  type: z.string(),
+  jobId: z.string(),
+  date: z.string(),
+  duration: z.string().optional(),
+  totalDuration: z.string().optional(),
+  workerId: z.string(),
+  attempt: z.number(),
+  context: context.nullable(),
+});
+
+export default WebhookEndpoint(async (req, res) => {
+  if (req.method !== 'POST') {
+    return res.status(405).json({ error: 'Method not allowed' });
+  }
+
+  const bodyResults = schema.safeParse(req.body);
+  if (!bodyResults.success) {
+    return res.status(400).json({ ok: false, errors: bodyResults.error });
+  }
+
+  const data = bodyResults.data;
+
+  const status = {
+    Success: TrainingStatus.InReview,
+    Fail: TrainingStatus.Failed,
+    Update: TrainingStatus.Processing,
+  }[data.type];
+
+  switch (data.type) {
+    case 'Success':
+    case 'Fail':
+    case 'Update':
+      // TODO: this validation could be done by zod but would need to handle all the other type cases
+      if (!data.context) {
+        return res.status(400).json({ ok: false, error: 'context is undefined' });
+      }
+
+      if (!data.context?.modelFileId) {
+        return res.status(400).json({ ok: false, error: 'modelFileId is undefined' });
+      }
+
+      await updateRecords(data.context, status as TrainingStatus);
+
+      break;
+    default:
+    // TODO: what should we do in the default case...
```

Comment:
throw an error and log to db using `logToDb`

Reply:
@rickbau5